### PR TITLE
Refactoring of Segment Replication classes

### PR DESCRIPTION
**Pull request by kartg**
### Description
Incremental refactoring of Segment Replication classes to centralize common functionality between seg-rep and recovery code. These are mostly modeled as `Replication*` parent classes with `Recovery*` and `SegmentReplication*` child classes.

There is much more refactoring that can be done above this, but the work in these commits is a time-boxed effort.
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [ ] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java [hunk 1]
```diff
@@ -748,7 +748,7 @@ private static DiscoveryNode findSourceNodeForPeerRecovery(
         return sourceNode;
     }
 
-    private class ReplicationListener implements SegmentReplicationReplicaService.ReplicationListener {
+    private class ShardRoutingReplicationListener implements SegmentReplicationReplicaService.SegmentReplicationListener {
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java:751]: This was me being quick and dirty with the POC - I don't think we need a separate listener type for segrep.  I think we could extract PeerRecoveryTargetService.RecoveryListener and use it for segrep with a common exception type.  We will still need to perform the same actions that `RecoveryListener` now performs on success (publishing shardStarted) and failure (failing and removing the shard).

File: server/src/main/java/org/opensearch/indices/replication/copy/ReplicationCollection.java [hunk 2]
```diff
@@ -35,7 +35,7 @@
 import org.apache.logging.log4j.Logger;
```
Comment (mch2) [inline server/src/main/java/org/opensearch/indices/replication/copy/ReplicationCollection.java:35]: Should we also rename this class to SegmentReplicationCollection?

File: server/src/main/java/org/opensearch/common/concurrent/AutoCloseableRefCounted.java [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+
+    private final RefCounted ref;
+    private final OneWayGate gate;
+
+    public AutoCloseableRefCounted(RefCounted ref) {
+        this.ref = ref;
+        gate = new OneWayGate();
+    }
+
+    public <T extends RefCounted> T get(Class<T> returnType) {
```
Comment (mch2) [inline server/src/main/java/org/opensearch/common/concurrent/AutoCloseableRefCounted.java:32]: I'm not in love with having to cast this every time.  Can we set the generic type on the class? `AutoCloseableRefCounted<T extends RefCounted>`?

Comment (kartg) [reply server/src/main/java/org/opensearch/common/concurrent/AutoCloseableRefCounted.java:32]: Good idea, done

File: server/src/main/java/org/opensearch/indices/replication/copy/ReplicationCollection.java (hunk 2 above)
Comment (kartg) [reply server/src/main/java/org/opensearch/indices/replication/copy/ReplicationCollection.java:35]: I haven't yet touched the Collection classes so i'd like to leave this as-is for now. In a follow up refactor, we can rename this appropriately after synergizing functionality with the Recovery equivalent

File: server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java (hunk 1 above)
Comment (kartg) [reply server/src/main/java/org/opensearch/indices/cluster/IndicesClusterStateService.java:751]: Understood. I'll punt this to a future refactor

**Review by mch2 — APPROVED** [review]
lgtm, thanks!

**Pull request merged**